File: ois/ois20/round2/ois11_round2_solutions/bonus.c
/*
 * This template is valid both in C and in C++,
 * so you can expand it with code from both languages.
 * NOTE: it is recommended to use this even if you don't
 * understand the following code.
 */

#include <stdio.h>
#include <assert.h>

// constraints
#define MAXN 100000

// input data
long long N, i;
long long S[MAXN], P[MAXN];

int main() {
//  uncomment the following lines if you want to read/write from files
//  freopen("input.txt", "r", stdin);
//  freopen("output.txt", "w", stdout);

    assert(1 == scanf("%lld", &N));
    for(i=0; i<N; i++) {
        assert(1 == scanf("%lld", &S[i]));
        assert(1 == scanf("%lld", &P[i]));
    }

    long long sumP = 0;
    long long sumS = 0;
    for (i = 0; i < N; i++) {
        sumP += P[i];
        sumS += S[i];
    }

    long long best = 0;
    for (i = 0; i < N; i++) {
        if ((sumS - S[i]) * (sumP - P[best]) >= (sumS - S[best]) * (sumP - P[i])) {
            best = i;
        }
    }

    printf("%lld\n", best); // print the result
    return 0;
}

File: ois/ois20/round2/ois11_round2_solutions/bonus.pas
{ NOTE: it is recommended to use this even if you don't understand the following code. }

{ constraints }
const
    MAXN = 100000;

{ input data }
var
    N, i  : Int64;
    S, P  : array[0..MAXN-1] of Int64;
    best  : Int64;
    sumS, sumP: Int64;
begin
{
    uncomment the following lines if you want to read/write from files
    assign(input,  'input.txt');  reset(input);
    assign(output, 'output.txt'); rewrite(output);
}

    readln(N);
    sumP := 0;
    sumS := 0;
    for i:=0 to N-1 do
    begin
        read(S[i]);
        read(P[i]);
        readln();

        sumP := sumP + P[i];
        sumS := sumS + S[i];
    end;

    best := 0;
    for i:=0 to N-1 do
    begin
        if ((sumS - S[i]) * (sumP - P[best]) > (sumS - S[best]) * (sumP - P[i])) then
            best := i
    end;

    writeln(best); { print result }
end.
